### PR TITLE
TRPC Movies and Cinema

### PR DESCRIPTION
**Pull request by MuliaHartawan**
### Entity Relationship Diagram
EM Manage
### What did you do?
1. added gitignore for path `public/upload/`
2. changed prisma provider to mysql
3. create schema and model movie, genre, movie-genre, studio, screening-schedule
4. create seeder
5. create api to upload image
6. create TRPC routes from each domain (movie, genre, movie-genre, studio, screening-schedule)
7. create validation using zod according to the desired data type 
8. create an action to check the validation of the TRPC input pubic procedure
9. create repository from each domain (movie, genre, movie-genre, studio, screening-schedule)

### CODE REVIEW

File: src/libs/attachment/image.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    const fileExtension = path.extname(file.name);
+    const uniqueFileName = `${uuidv4()}${fileExtension}`;
+    const filePath = path.join(UPLOAD_DIR, uniqueFileName);
+
+    const bytes = await file.arrayBuffer();
+    const buffer = Buffer.from(bytes);
+
+    await fs.writeFile(filePath, buffer);
```
Comment (dprastha) [inline src/libs/attachment/image.ts]: - Proses validasi dan write file bisa dijadikan util mas @MuliaHartawan 
- Tambahkan validasi max size dan tipe file misal .png dan .jpg

File: src/server/screen-schedule/validations/create-or-update-screen-schedule.validation.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { z } from "zod";
+
+export const createOrUpdateScreenScheduleSchema = z.object({
+  screeningTime: z
+    .string({
+      message: "screening Tim harus berupa string",
+    })
+    .regex(/^\d{4}-\d{2}-\d{2}$/, { message: "Format release date harus YYYY-MM-DD" }),
```
Comment (egagofur) [inline src/server/screen-schedule/validations/create-or-update-screen-schedule.validation.ts]: mungkin bisa di simpan sebagai constant karena dipakai buat kebutuhan format date mas

File: src/server/movie/actions/movie.action.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  await updateMovieAndGenres(id, {
+    title: value.title,
+    releaseDate: new Date(value.releaseDate),
+    duration: value.duration,
+    description: value.description ?? null,
+    rating: value.rating,
+    poster: value.poster,
+    genreIds: value.genreIds,
+  });
+};
```
Comment (dprastha) [inline src/server/movie/actions/movie.action.ts]: Update data tidak perlu find data terlebih dahulu mas @MuliaHartawan

File: src/server/movie/validations/create-or-update-movie.validation.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  title: z
+    .string({ required_error: "Title Wajib Diisi", message: "Title harus berupa string" })
+    .min(1, { message: "Title wajib diisi" })
+    .max(255, { message: "Title maksimal 255 karakter" }),
+
+  releaseDate: z
+    .string({
+      message: "Release date harus berupa string",
+    })
+    .regex(/^\d{4}-\d{2}-\d{2}$/, { message: "Format release date harus YYYY-MM-DD" }),
```
Comment (dprastha) [inline src/server/movie/validations/create-or-update-movie.validation.ts]: bisa langsung pakai tipe date mas @MuliaHartawan ?

File: src/server/movie/actions/movie.action.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  // Permission authorization
+  await serverCheckPermission([PERMISSIONS.MOVIE_DELETE]);
+
+  const movie = await findOneMovieById(id);
+  if (!movie) {
+    throw new NotFoundException("Movie tidak ditemukan!");
+  }
+
+  await deleteMovieById(id);
+};
```
Comment (dprastha) [inline src/server/movie/actions/movie.action.ts]: Idem https://github.com/MuliaHartawan/movie-nextjs-trpc/pull/6/files#r1875406433

File: src/server/screen-schedule/actions/screen-schedule.action.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  if (!screenSchedule) {
+    throw new NotFoundException("Screen schedule tidak ditemukan!");
+  }
+
+  await updateScreenScheduleAndGenres(id, {
+    screeningTime: new Date(value.screeningTime),
+    price: value.price,
+    movieId: value.movieId,
+    studioId: value.studioId,
+  });
```
Comment (dprastha) [inline src/server/screen-schedule/actions/screen-schedule.action.ts]: Idem mas @MuliaHartawan https://github.com/MuliaHartawan/movie-nextjs-trpc/pull/6/files#r1875406433

File: src/server/screen-schedule/actions/screen-schedule.action.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+export const deleteScreenScheduleAction = async (id: string) => {
+  // Permission authorization
+  await serverCheckPermission([PERMISSIONS.SCREEN_SCHEDULE_DELETE]);
+
+  const screenSchedule = await findOneScreenScheduleById(id);
+  if (!screenSchedule) {
+    throw new NotFoundException("Screen schedule tidak ditemukan!");
+  }
```
Comment (dprastha) [inline src/server/screen-schedule/actions/screen-schedule.action.ts]: https://github.com/MuliaHartawan/movie-nextjs-trpc/pull/6/files#r1875406433

File: src/server/studio/actions/studio.action.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  // Permission authorization
+  await serverCheckPermission([PERMISSIONS.STUDIO_UPDATE]);
+
+  // Validation
+  await validate(createOrUpdateStudioSchema, value);
+
+  const studio = await findOneStudioById(id);
+  if (!studio) {
+    throw new NotFoundException("Sudio tidak ditemukan!");
+  }
```
Comment (dprastha) [inline src/server/studio/actions/studio.action.ts]: idem https://github.com/MuliaHartawan/movie-nextjs-trpc/pull/6/files#r1875406433

File: src/server/studio/actions/studio.action.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+export const deleteStudioAction = async (id: string) => {
+  // Permission authorization
+  await serverCheckPermission([PERMISSIONS.STUDIO_DELETE]);
+
+  const studio = await findOneStudioById(id);
+  if (!studio) {
+    throw new NotFoundException("Sudio tidak ditemukan!");
+  }
+
+  await deleteStudioById(id);
```
Comment (dprastha) [inline src/server/studio/actions/studio.action.ts]: idem https://github.com/MuliaHartawan/movie-nextjs-trpc/pull/6/files#r1875406433

File: src/server/movie/validations/create-or-update-movie.validation.ts (hunk 4 above)
Comment (egagofur) [inline src/server/movie/validations/create-or-update-movie.validation.ts]: mungkin bisa dipindahkan ke constant mas

File: src/libs/attachment/image.ts (hunk 1 above)
Comment (MuliaHartawan) [reply src/libs/attachment/image.ts]: sudah aku refactor codenya mas untuk membuat file validation di utils bisa cek kembali mas @dprastha

File: src/server/movie/actions/movie.action.ts (hunk 3 above)
Comment (MuliaHartawan) [reply src/server/movie/actions/movie.action.ts]: kenapa seperti mas @dprastha ? bukannya harus di pastikan dulu apakah datanya existing?

Comment (MuliaHartawan) [reply src/server/movie/actions/movie.action.ts]: ini sudah aku remove mas @dprastha

File: src/server/screen-schedule/actions/screen-schedule.action.ts (hunk 6 above)
Comment (MuliaHartawan) [reply src/server/screen-schedule/actions/screen-schedule.action.ts]: ini sudah aku remove mas @dprastha

File: src/server/studio/actions/studio.action.ts (hunk 8 above)
Comment (MuliaHartawan) [reply src/server/studio/actions/studio.action.ts]: ini sudah aku remove mas @dprastha

File: src/server/studio/actions/studio.action.ts (hunk 9 above)
Comment (MuliaHartawan) [reply src/server/studio/actions/studio.action.ts]: ini sudah aku remove mas @dprastha

File: src/server/screen-schedule/actions/screen-schedule.action.ts (hunk 7 above)
Comment (MuliaHartawan) [reply src/server/screen-schedule/actions/screen-schedule.action.ts]: ini sudah aku remove mas @dprastha

File: src/server/movie/actions/movie.action.ts (hunk 5 above)
Comment (MuliaHartawan) [reply src/server/movie/actions/movie.action.ts]: ini sudah aku remove mas @dprastha

**Pull request merged**